### PR TITLE
Use VectorFuzzer::fuzzInputFlatRow in ParquetWriterTestBase

### PR DESCRIPTION
**Pull request by duanmeng**
It is a good practice to use fuzzer to generate test data.

Update ParquetWriterTestBase to use VectorFuzzer::fuzzInputFlatRow to generate 
all flat vectors suitable for writing using Arrow Parquet writer. Note that Arrow Parquet 
writer doesn't support encoded data.

### CODE REVIEW

**Review by xiaoxmeng — APPROVED** [review]
@duanmeng LGTM. Thanks for the change!

**Comment by mbasmanova** [comment]
@duanmeng I wonder why can't we use fuzzFlat method?

File: velox/vector/fuzzer/tests/VectorFuzzerTest.cpp [hunk 1]
```diff
@@ -688,6 +688,29 @@ TEST_F(VectorFuzzerTest, fuzzRowChildrenToLazy) {
...
+  opts.dictionaryRatio = 0.0;
+  opts.vectorSize = 100;
+  VectorFuzzer fuzzer(opts, pool());
+
+  bool hasConstant = false;
+  bool hasDictionary = false;
+  for (size_t i = 0; i < 100; ++i) {
+    auto type = fuzzer.randType();
+    auto vector = fuzzer.fuzz(type);
+    if (vector->encoding() == VectorEncoding::Simple::CONSTANT) {
```
Comment (mbasmanova) [inline velox/vector/fuzzer/tests/VectorFuzzerTest.cpp:703]: Why not ASSERT_NE(vector->encoding(), VectorEncoding::Simple::CONSTANT)

File: velox/vector/fuzzer/tests/VectorFuzzerTest.cpp [hunk 2]
```diff
@@ -688,6 +688,29 @@ TEST_F(VectorFuzzerTest, fuzzRowChildrenToLazy) {
...
+      hasConstant = true;
+    }
+
+    if (vector->encoding() == VectorEncoding::Simple::DICTIONARY) {
+      hasDictionary = true;
```
Comment (mbasmanova) [inline velox/vector/fuzzer/tests/VectorFuzzerTest.cpp:708]: This variable is not used.

File: velox/vector/fuzzer/tests/VectorFuzzerTest.cpp (hunk 1 above)
Comment (duanmeng) [reply velox/vector/fuzzer/tests/VectorFuzzerTest.cpp]: Fixed

File: velox/vector/fuzzer/tests/VectorFuzzerTest.cpp (hunk 2 above)
Comment (duanmeng) [reply velox/vector/fuzzer/tests/VectorFuzzerTest.cpp]: Fixed

**Comment by duanmeng** [comment]
> @duanmeng I wonder why can't we use fuzzFlat method?

@mbasmanova Hi Masha, I used to want to create a RowVector batch in this [PR](https://github.com/facebookincubator/velox/pull/6608/files#r1337578838) like here in [HiveDataSinkTest](https://github.com/facebookincubator/velox/blob/main/velox/connectors/hive/tests/HiveDataSinkTest.cpp#L409). `fuzzFlat` maybe not convenient since I want to generate a row with multiple columns.

File: velox/vector/fuzzer/tests/VectorFuzzerTest.cpp [hunk 3]
```diff
@@ -688,6 +688,24 @@ TEST_F(VectorFuzzerTest, fuzzRowChildrenToLazy) {
...
 }
 
+TEST_F(VectorFuzzerTest, constantOpt) {
+  VectorFuzzer::Options opts;
+  opts.constantRatio = 0.0;
+  opts.dictionaryRatio = 0.0;
+  opts.vectorSize = 100;
+  VectorFuzzer fuzzer(opts, pool());
+
+  bool hasConstant = false;
```
Comment (mbasmanova) [inline velox/vector/fuzzer/tests/VectorFuzzerTest.cpp:698]: These variables not used.

File: velox/vector/fuzzer/tests/VectorFuzzerTest.cpp [hunk 4]
```diff
@@ -688,6 +688,24 @@ TEST_F(VectorFuzzerTest, fuzzRowChildrenToLazy) {
   ASSERT_TRUE(wrappedRow->childAt(1)->as<LazyVector>()->isLoaded());
 }
 
+TEST_F(VectorFuzzerTest, constantOpt) {
```
Comment (mbasmanova) [inline velox/vector/fuzzer/tests/VectorFuzzerTest.cpp:691]: naming: do not abbreviate: constantRatio

However, this test tests both constant and dictionary ratio, hence the name doesn't reflect the test logic. It might be encodingRatios, but then the test only tests zero ratios. You may need to extend the test to verify multiple combinations: both ratios zero, only one ratio is zero but < 1, both non-zero but < 1, both are 1, one is 1, etc.

File: velox/dwio/parquet/tests/ParquetWriterTestBase.h [hunk 5]
```diff
@@ -35,17 +36,23 @@ class ParquetWriterTestBase : public testing::Test {
...
+        ROW({"c0", "c1", "c2", "c3", "c4", "c5"},
+            {BIGINT(), INTEGER(), SMALLINT(), REAL(), DOUBLE(), VARCHAR()});
   }
 
   std::vector<RowVectorPtr> createBatches(
-      const std::shared_ptr<const Type>& rowType,
       uint64_t batchNum,
-      uint64_t capacity) {
+      uint64_t vectorSize) {
+    VectorFuzzer::Options opts;
```
Comment (mbasmanova) [inline velox/dwio/parquet/tests/ParquetWriterTestBase.h:47]: Notice that this logic will create same exact batch every time. Is this the intention?

File: velox/dwio/parquet/tests/ParquetWriterTestBase.h [hunk 6]
```diff
@@ -35,17 +36,23 @@ class ParquetWriterTestBase : public testing::Test {
...
+    opts.constantRatio = 0.0;
+    opts.dictionaryRatio = 0.0;
```
Comment (mbasmanova) [inline velox/dwio/parquet/tests/ParquetWriterTestBase.h:49]: I feel there are so many options in the Fuzzer already that it becomes difficult to reason about what sort of randomness is provided in any particular use case. I wonder if it would just simpler and more readable to use fuzzFlat.

```
for (auto i = 0; i < rowType_->size(); ++i) {
   children.push_back(fuzzer.fuzzFlat(rowType_->childAt(i));
}
makeRowVector(children);
```

Or, add fuzzFlatInputRow API.

File: velox/dwio/parquet/tests/ParquetWriterTestBase.h [hunk 7]
```diff
@@ -24,6 +24,7 @@
 #include "velox/dwio/common/tests/utils/DataFiles.h"
 #include "velox/dwio/parquet/writer/Writer.h"
 #include "velox/exec/tests/utils/TempDirectoryPath.h"
+#include "velox/vector/fuzzer/VectorFuzzer.h"
```
Comment (mbasmanova) [inline velox/dwio/parquet/tests/ParquetWriterTestBase.h:27]: Can we now remove 'include' for BatchMaker?

File: velox/dwio/parquet/tests/ParquetWriterTestBase.h [hunk 8]
```diff
@@ -35,17 +36,23 @@ class ParquetWriterTestBase : public testing::Test {
     rootPool_ = memory::defaultMemoryManager().addRootPool("ParquetWriterTest");
     leafPool_ = rootPool_->addLeafChild("SinkTest");
     tempPath_ = exec::test::TempDirectoryPath::create();
+    rowType_ =
```
Comment (mbasmanova) [inline velox/dwio/parquet/tests/ParquetWriterTestBase.h:39]: Move initialization to declaration.

File: velox/dwio/parquet/tests/ParquetWriterTestBase.h (hunk 6 above)
Comment (duanmeng) [reply velox/dwio/parquet/tests/ParquetWriterTestBase.h:49]: >add fuzzFlatInputRow API

Got it, I will use this solution.

Comment (xiaoxmeng) [reply velox/dwio/parquet/tests/ParquetWriterTestBase.h:49]: @mbasmanova I were wondering how to convert a vector shared ptr to row vector shared ptr :)

Comment (mbasmanova) [reply velox/dwio/parquet/tests/ParquetWriterTestBase.h:49]: std::dynamic_pointer_cast

File: velox/dwio/parquet/tests/ParquetWriterTestBase.h (hunk 5 above)
Comment (duanmeng) [reply velox/dwio/parquet/tests/ParquetWriterTestBase.h]: I wanted to use it to generate some data to write no matter what the content is.

File: velox/dwio/parquet/tests/ParquetWriterTestBase.h (hunk 7 above)
Comment (duanmeng) [reply velox/dwio/parquet/tests/ParquetWriterTestBase.h:26]: Removed

File: velox/dwio/parquet/tests/ParquetWriterTestBase.h (hunk 8 above)
Comment (duanmeng) [reply velox/dwio/parquet/tests/ParquetWriterTestBase.h]: Fixed

File: velox/vector/fuzzer/tests/VectorFuzzerTest.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (duanmeng) [reply velox/vector/fuzzer/tests/VectorFuzzerTest.cpp]: I've updated the PR, added a fuzzInputFlatRow API, and removed the new opt and this ut.

Comment (duanmeng) [reply velox/vector/fuzzer/tests/VectorFuzzerTest.cpp]: I've updated the PR, added a fuzzInputFlatRow API, and removed the new opt and this ut.

File: velox/dwio/parquet/tests/ParquetWriterTestBase.h [hunk 9]
```diff
@@ -20,32 +20,29 @@
...
     leafPool_ = rootPool_->addLeafChild("SinkTest");
     tempPath_ = exec::test::TempDirectoryPath::create();
   }
 
-  std::vector<RowVectorPtr> createBatches(
-      const std::shared_ptr<const Type>& rowType,
-      uint64_t batchNum,
-      uint64_t capacity) {
+  std::vector<RowVectorPtr>
+  createBatches(uint64_t batchNum, uint64_t vectorSize, RowTypePtr rowType) {
```
Comment (mbasmanova) [inline velox/dwio/parquet/tests/ParquetWriterTestBase.h:40]: Would you please move the implementation into .cpp time to speed up incremental builds?

Comment (mbasmanova) [inline velox/dwio/parquet/tests/ParquetWriterTestBase.h:40]: batchNum -> numBatches

pass rowType as const &

File: velox/dwio/parquet/tests/ParquetWriterTestBase.h [hunk 10]
```diff
@@ -69,8 +66,14 @@ class ParquetWriterTestBase : public testing::Test {
     return std::make_unique<Writer>(std::move(sink), options);
   }
 
-  std::shared_ptr<memory::MemoryPool> rootPool_;
-  std::shared_ptr<memory::MemoryPool> leafPool_;
-  std::shared_ptr<exec::test::TempDirectoryPath> tempPath_;
+  static std::shared_ptr<memory::MemoryPool> rootPool_;
```
Comment (mbasmanova) [inline velox/dwio/parquet/tests/ParquetWriterTestBase.h:69]: Why static?

File: velox/vector/fuzzer/VectorFuzzer.h [hunk 11]
```diff
@@ -238,6 +238,10 @@ class VectorFuzzer {
   // elements.
   RowVectorPtr fuzzInputRow(const RowTypePtr& rowType);
 
+  // Returns a RowVector whose children are flat vectors or a complex vectors
```
Comment (mbasmanova) [inline velox/vector/fuzzer/VectorFuzzer.h:241]: Use ///

Comment (mbasmanova) [inline velox/vector/fuzzer/VectorFuzzer.h:241]: Would you extract fuzzer changes into a separate PR?

Comment (duanmeng) [reply velox/vector/fuzzer/VectorFuzzer.h:241]: @mbasmanova I've filed another PR https://github.com/facebookincubator/velox/pull/6849 extracting the fuzzer changes. I will continue to use this PR to refactor `ParquetWriterTestBase.h`.

**Comment by duanmeng** [comment]
@mbasmanova Hi Masha, I've updated this PR to use the new `fuzzerInputFlatRow` API and move the implementation into .cpp time to speed up incremental builds. Could you please help to take a look?

File: velox/dwio/parquet/tests/ParquetWriterTestBase.h (hunk 7 above)
Comment (mbasmanova) [reply velox/dwio/parquet/tests/ParquetWriterTestBase.h:27]: @duanmeng For some reason, I still see

#include "velox/dwio/common/tests/utils/BatchMaker.h"

File: velox/dwio/parquet/tests/writer/CMakeLists.txt [hunk 12]
```diff
@@ -12,7 +12,8 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-add_executable(velox_parquet_writer_sink_test SinkTests.cpp)
+add_executable(velox_parquet_writer_sink_test SinkTests.cpp
+                                              ../ParquetWriterTestBase.cpp)
```
Comment (mbasmanova) [inline velox/dwio/parquet/tests/writer/CMakeLists.txt:16]: Hmm.... why do we include a file from the parent directory? This is an anti-pattern.

File: velox/dwio/parquet/tests/ParquetWriterTestBase.cpp [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+
+namespace facebook::velox::parquet {
+
+std::vector<RowVectorPtr> ParquetWriterTestBase::createBatches(
+    const RowTypePtr& rowType,
+    uint64_t numBatches,
+    uint64_t vectorSize) {
+  std::vector<RowVectorPtr> batches;
+  for (auto i = 0; i < numBatches; ++i) {
+    VectorFuzzer fuzzer({.vectorSize = vectorSize}, leafPool_.get());
```
Comment (mbasmanova) [inline velox/dwio/parquet/tests/ParquetWriterTestBase.cpp:27]: Do we need to re-create fuzzer for each batch? Why can't we re-use fuzzer for all batches?

File: velox/dwio/parquet/tests/ParquetWriterTestBase.cpp [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+    fuzzer.fuzzInputFlatRow(rowType);
```
Comment (mbasmanova) [inline velox/dwio/parquet/tests/ParquetWriterTestBase.cpp:28]: Why call fuzzInputFlatRow and drop the results?

**Review by mbasmanova — COMMENTED** [review]
Some comments. Please, update PR description to match the changes.

File: velox/dwio/parquet/tests/writer/CMakeLists.txt (hunk 12 above)
Comment (duanmeng) [reply velox/dwio/parquet/tests/writer/CMakeLists.txt]: Fixed by moving them to the test/writer

File: velox/dwio/parquet/tests/ParquetWriterTestBase.cpp (hunk 13 above)
Comment (duanmeng) [reply velox/dwio/parquet/tests/ParquetWriterTestBase.cpp]: Fixed

File: velox/dwio/parquet/tests/ParquetWriterTestBase.cpp (hunk 14 above)
Comment (duanmeng) [reply velox/dwio/parquet/tests/ParquetWriterTestBase.cpp]: Fixed

**Comment by duanmeng** [comment]
> Some comments. Please, update PR description to match the changes.

@mbasmanova Got it, updated the description and resolved the comments.

File: velox/dwio/parquet/tests/CMakeLists.txt [hunk 15]
```diff
@@ -15,6 +15,7 @@
 set(TEST_LINK_LIBS
     velox_dwio_common_test_utils
     velox_vector_test_lib
+    velox_vector_fuzzer
```
Comment (mbasmanova) [inline velox/dwio/parquet/tests/CMakeLists.txt:18]: This should go into the other cmake file.

File: velox/dwio/parquet/tests/writer/SinkTests.cpp [hunk 16]
```diff
@@ -16,9 +16,8 @@
 
 #include <utility>
 
+#include "ParquetWriterTestBase.h"
```
Comment (mbasmanova) [inline velox/dwio/parquet/tests/writer/SinkTests.cpp:19]: Use full paths: velox/...

File: velox/dwio/parquet/tests/writer/ParquetWriterTestBase.h [hunk 17]
```diff
@@ -20,11 +20,10 @@
 #include <string>
 #include "velox/common/base/Fs.h"
 #include "velox/dwio/common/FileSink.h"
-#include "velox/dwio/common/tests/utils/BatchMaker.h"
 #include "velox/dwio/common/tests/utils/DataFiles.h"
 #include "velox/dwio/parquet/writer/Writer.h"
 #include "velox/exec/tests/utils/TempDirectoryPath.h"
```
Comment (mbasmanova) [inline velox/dwio/parquet/tests/writer/ParquetWriterTestBase.h:25]: Now that you moved implementation to .cpp file. Do you still need all these includes?

File: velox/dwio/parquet/tests/writer/ParquetWriterTestBase.cpp [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+
+namespace facebook::velox::parquet {
+
+std::vector<RowVectorPtr> ParquetWriterTestBase::createBatches(
+    const RowTypePtr& rowType,
+    uint64_t numBatches,
+    uint64_t vectorSize) {
+  std::vector<RowVectorPtr> batches;
+  batches.reserve(numBatches);
+  VectorFuzzer fuzzer({.vectorSize = vectorSize}, leafPool_.get());
```
Comment (mbasmanova) [inline velox/dwio/parquet/tests/writer/ParquetWriterTestBase.cpp:27]: By default, fuzzer doesn't generate nulls. Is this the desired behavior here?

File: velox/dwio/parquet/tests/writer/ParquetWriterTestBase.cpp [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+std::unique_ptr<Writer> ParquetWriterTestBase::createWriter(
+    std::unique_ptr<dwio::common::FileSink> sink,
+    std::function<std::unique_ptr<DefaultFlushPolicy>()> flushPolicy) {
+  facebook::velox::parquet::WriterOptions options;
+  options.memoryPool = rootPool_.get();
+  options.flushPolicyFactory = flushPolicy;
+  return std::make_unique<Writer>(std::move(sink), options);
+}
+
+} // namespace facebook::velox::parquet
```
Comment (mbasmanova) [inline velox/dwio/parquet/tests/writer/ParquetWriterTestBase.cpp:53]: add newline at the end of the file

File: velox/dwio/parquet/tests/CMakeLists.txt (hunk 15 above)
Comment (duanmeng) [reply velox/dwio/parquet/tests/CMakeLists.txt]: Fixed

File: velox/dwio/parquet/tests/writer/SinkTests.cpp (hunk 16 above)
Comment (duanmeng) [reply velox/dwio/parquet/tests/writer/SinkTests.cpp]: Fixed, sorry that I did not notice here.

File: velox/dwio/parquet/tests/writer/ParquetWriterTestBase.h (hunk 17 above)
Comment (duanmeng) [reply velox/dwio/parquet/tests/writer/ParquetWriterTestBase.h:23]: Fixed by move some includes into cpp file.

File: velox/dwio/parquet/tests/writer/ParquetWriterTestBase.cpp (hunk 19 above)
Comment (duanmeng) [reply velox/dwio/parquet/tests/writer/ParquetWriterTestBase.cpp]: Fixed

File: velox/dwio/parquet/tests/writer/ParquetWriterTestBase.cpp (hunk 18 above)
Comment (duanmeng) [reply velox/dwio/parquet/tests/writer/ParquetWriterTestBase.cpp:29]: We just need some data to write here to check close/abort here.

**Comment by majetideepak** [comment]
@mbasmanova: @czentgr, and I just looked into this yesterday. Christian observed that the dictionary encoding generated by partition column tests in the TableWriteTest.cpp does not work for Arrow Parquet. (This issue is being tracked here https://github.com/facebookincubator/velox/issues/5560). 
There are two issues.
1) The indices buffer is reused across batches and this broke the writer when chunk staging was introduced here  https://github.com/facebookincubator/velox/pull/5442.
2) Arrow Parquet also does not support writing Null dictionary values.

So we need to flatten the dictionary values before passing the vector to the Arrow Bridge to fix these issues.

File: velox/dwio/parquet/tests/writer/ParquetWriterTestBase.cpp (hunk 18 above)
Comment (majetideepak) [reply velox/dwio/parquet/tests/writer/ParquetWriterTestBase.cpp:29]: @duanmeng we need nulls for other tests.
@mbasmanova How do we currently generate nulls along with the Fuzzer for other tests?

Comment (mbasmanova) [reply velox/dwio/parquet/tests/writer/ParquetWriterTestBase.cpp:29]: @majetideepak Fuzzer allows to specify a fraction of nulls:

```
class VectorFuzzer {
 public:
  struct Options {
    size_t vectorSize{100};

    /// Chance of generating a null value in the output vector (`nullRatio` is a
    /// double between 0 and 1).
    double nullRatio{0};
```
